docs/architecture/adr-d001-inter-validator-set-messaging.md

```diff
@@ -0,0 +1,243 @@
+   1. [Definitions](#definitions)
+   2. [Problem statement](#problem-statement)
+2. [Alternative Approaches](#alternative-approaches)
+   1. [Scenario 1: Trigger rotation from ABCI app passing addresses of all quorum members](#scenario-1-trigger-rotation-from-abci-app-passing-addresses-of-all-quorum-members)
+   2. [Scenario 2: Trigger rotation from ABCI app passing quorum id to rotate](#scenario-2-trigger-rotation-from-abci-app-passing-quorum-id-to-rotate)
+   3. [Scenario 3: Rotate on Tenderdash](#scenario-3-rotate-on-tenderdash)
+3. [Decision](#decision)
+4. [Detailed Design](#detailed-design)
+   1. [What are the user requirements?](#what-are-the-user-requirements)
+   2. [What systems will be affected?](#what-systems-will-be-affected)
+   3. [What new data structures are needed, what data structures need changes?](#what-new-data-structures-are-needed-what-data-structures-need-changes)
+   4. [What new APIs will be needed, what APIs will change?](#what-new-apis-will-be-needed-what-apis-will-change)
+   5. [What are the efficiency considerations (time/space)?](#what-are-the-efficiency-considerations-timespace)
+   6. [What are the expected access patterns (load/throughput)?](#what-are-the-expected-access-patterns-loadthroughput)
+   7. [Are there any logging, monitoring, or observability needs?](#are-there-any-logging-monitoring-or-observability-needs)
+   8. [Are there any security considerations?](#are-there-any-security-considerations)
+   9. [Are there any privacy considerations?](#are-there-any-privacy-considerations)
+   10. [How will the changes be tested?](#how-will-the-changes-be-tested)
+   11. [How will the changes be broken up for ease of review?](#how-will-the-changes-be-broken-up-for-ease-of-review)
+   12. [Will these changes require a breaking (major) release?](#will-these-changes-require-a-breaking-major-release)
+   13. [Does this change require coordination with the SDK or any other software?](#does-this-change-require-coordination-with-the-sdk-or-any-other-software)
+5. [Status](#status)
+6. [Consequences](#consequences)
+   1. [Positive](#positive)
+   2. [Negative](#negative)
+   3. [Neutral](#neutral)
+7. [References](#references)
+
+## Context
+
+### Definitions
+
+For the needs of this document, we define the following types of nodes:
+
+* Validator nodes, which participate in the consensus protocol, verify and sign blocks, 
+* Full nodes, which execute valid blocks.
+
+Note: there are also light nodes, but light nodes are not relevant for the discussion.
+
+Validator Set is a group of Validator nodes responsible for the consensus at a given time. Only one Validator Set can
+be active at the given time. After a predefined number of blocks, the ABCI app initiates the Validator Set rotation to
+make another Validator Set active. 
+
+### Problem statement
+
+The consensus process requires direct communication between validators, as full nodes do not accept nor propagate
+consensus protocol messages (like votes).
```

Comment:
The consensus process requires direct communication between validators, as full nodes do not know the public keys of validators and therefore can neither verify nor propagate any consensus protocol message that are individually signed by these validators. This means that votes can only be propagated by validators. Full nodes can however propagate the final commit message, which instead contains the recovered threshold signature of the precommit voting round. Full nodes can perform this verification because they know the validator set's quorum threshold public key.